nifi-commons/nifi-record/src/test/java/org/apache/nifi/serialization/record/ResultSetRecordSetTest.java

```diff
@@ -272,6 +278,76 @@ public void testCreateSchemaArrayThrowsNotSupportedException() throws SQLExcepti
     }
 
+    @Test
+    public void testCreateSchemaWithLogicalTypes() throws SQLException {
+        testCreateSchemaLogicalTypes(true);
+    }
+
+    @Test
+    public void testCreateSchemaNoLogicalTypes() throws SQLException {
+        testCreateSchemaLogicalTypes(false);
+    }
+
+    private void testCreateSchemaLogicalTypes(boolean useLogicalTypes) throws SQLException {
+        // GIVEN
+        Object[][] columns = new Object[][] {
+                {1, COLUMN_NAME_DATE, Types.DATE, RecordFieldType.DATE.getDataType()},
+                {2, "time", Types.TIME, RecordFieldType.TIME.getDataType()},
+                {3, "time_with_timezone", Types.TIME_WITH_TIMEZONE, RecordFieldType.TIME.getDataType()},
+                {4, "timestamp", Types.TIMESTAMP, RecordFieldType.TIMESTAMP.getDataType()},
+                {5, "timestamp_with_timezone", Types.TIMESTAMP_WITH_TIMEZONE, RecordFieldType.TIMESTAMP.getDataType()},
+                {6, COLUMN_NAME_BIG_DECIMAL_1, Types.DECIMAL,RecordFieldType.DECIMAL.getDecimalDataType(7, 3)},
+                {7, COLUMN_NAME_BIG_DECIMAL_2, Types.NUMERIC, RecordFieldType.DECIMAL.getDecimalDataType(4, 0)},
+                {8, COLUMN_NAME_BIG_DECIMAL_3, Types.JAVA_OBJECT, RecordFieldType.DECIMAL.getDecimalDataType(501, 1)},
+                {9, COLUMN_NAME_BIG_DECIMAL_4, Types.DECIMAL, RecordFieldType.DECIMAL.getDecimalDataType(10, 3)},
+                {10, COLUMN_NAME_BIG_DECIMAL_5, Types.DECIMAL, RecordFieldType.DECIMAL.getDecimalDataType(3, 10)}
+        };
+        final RecordSchema recordSchema = givenRecordSchema(columns);
+
+        ResultSetMetaData resultSetMetaData = Mockito.mock(ResultSetMetaData.class);
+        ResultSet resultSet = Mockito.mock(ResultSet.class);
+
+        RecordSchema expectedSchema = useLogicalTypes ? givenRecordSchema(columns) : givenRecordSchemaWithOnlyStringType(columns);
+
+        // WHEN
+        setUpMocks(columns, resultSetMetaData, resultSet);
+
+        ResultSetRecordSet testSubject = new ResultSetRecordSet(resultSet, recordSchema, 10,0, useLogicalTypes);
+        RecordSchema actualSchema = testSubject.getSchema();
+
+        // THEN
+        thenAllColumnDataTypesAreCorrect(columns, expectedSchema, actualSchema);
+    }
+
+    private void setUpMocks(Object[][] columns, ResultSetMetaData resultSetMetaData, ResultSet resultSet) throws SQLException {
+        when(resultSet.getMetaData()).thenReturn(resultSetMetaData);
+        when(resultSetMetaData.getColumnCount()).thenReturn(columns.length);
+
+        int indexOfBigDecimal = -1;
+        int index = 0;
+        for (final Object[] column : columns) {
+            when(resultSetMetaData.getColumnLabel((Integer) column[0])).thenReturn((String) column[1]);
+            when(resultSetMetaData.getColumnName((Integer) column[0])).thenReturn((String) column[1]);
+            when(resultSetMetaData.getColumnType((Integer) column[0])).thenReturn((Integer) column[2]);
```

Comment:
Restructuring the mocked columns would help avoid this type casting.

Reply:
The aforementioned TestColumn type will resolves this. Thank you!